src/main/java/org/jenkinsci/plugins/credentialsbinding/BindingDescriptor.java

```diff
@@ -44,6 +54,19 @@
 
     protected abstract Class<C> type();
 
+    /**
+     * Returns the context {@link MultiBinding} needs to access. Defaults to {@link Run}, {@link FilePath},
+     * {@link Launcher} and {@link TaskListener}.
+     *
+     * This allows the system to statically infer which steps are applicable in which context
+     * (say in freestyle or in workflow).
+     * @see StepContext#get(Class)
+     */
+    public Set<? extends Class<?>> getRequiredContext() {
```

Comment:
This does not really make sense as an API. There are two legitimate choices:

* `Run` + `TaskListener` + `FilePath` + `Launcher` (the default)
* `Run` + `TaskListener` (the new mode)

You could not usefully drop, say, `Run` or `TaskListener`—they will still be required. And you could not meaningfully request some additional context—you will indeed make the step fail if not called in that context, but you have no way of accessing the contextual object! (Since a `MultiBinding.bind` just gets fixed parameters.)

So I would just scrap this API and replace it with a simple

```java
public boolean requiresWorkspace() {
    return true;
}
```